### PR TITLE
chore: RFC for emitting multiple log events from remap

### PR DESCRIPTION
**Pull request by jszwedko**
RFC for https://github.com/timberio/vector/issues/6330

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+A new `emit_log` function will be added to the VRL stdlib:
+
+```text
+emit_log(value: Object)
+```
+
+This function will cause the object passed as value to be emitted at that point and flushed downstream. The emitted log will have its metadata copied from the input event.
+
+Additionally, an `emit_root` (we can work on the naming) config option will be added to the `remap` transform to configure whether `.` is emitted after the transform runs. It will default to `true` to preserve the current behavior but can be set to `false` by users to supress this behavior. Admittedly, I'm not wild about introducing this additional config option, but I'm not seeing another great alternative.
```
Comment (bruceg) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:35]: Could this be a flag set at runtime in the script?

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+This will be able to be combined with the iteration mechanism that will be introduced [#6031](https://github.com/timberio/vector/issues/6031) to emit an unknown number of events. Naively this might look something like:
+
+```text
+for stooge in .stooges
+   emit(stooge)
```
Comment (bruceg) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:41]: ITYM `emit_log`

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+end
+```
+
+In the future we can also add functions for emitting metrics like:
+
+```text
+emit_counter(namespace: String, name: String, timestamp: Timestamp, value: Float, kind: "absolute"|"relative")
+```
+
+I considered having just an `emit_metric()` but it would require users to pass in objects that match exactly the internal representation we have for metrics.
```
Comment (bruceg) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:51]: Not necessarily. The Lua wrapper has at various points converted to/from the internal format as necessary, as does the WASM wrapper. There's no reason this couldn't either.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+## Drawbacks
+
+- It makes the remap transform more complicated to understand and adds an additional configuration option that users need to be aware of
+- Ongoing maintenance burden should be minimal
+
+## Alternatives
+
+### Not emitting . if an `emit_log` is present in the program
+
+We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.
```
Comment (bruceg) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:83]: I was going to suggest this, but I agree it is a potentially surprising behavior. It also may introduce variable behavior if the script does call `emit_log` in some code path.

**Comment by FungusHumungus** [comment]
I'm hesitant to avoid the array option purely because of implementation difficulties. At some point we are going to want to change VRL so that it doesn't directly use the underlying event in order to avoid having to clone the object every time. 

If the array option feels better for the user, which I tend to lean towards, we could tackle both these issues together.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+https://github.com/timberio/vector/pull/6545
+
+Have a separate, special purpose, transform for converting a single log event into multiple log events.
+
+This would work, but I think the downsides are:
+
+- Introduces a new transform rather than just leveraging an existing one, remap. There is some precedence for this with the `route` and `reduce` transforms though.
+- It is less flexible than letting users emit arbitrary events
+- There isn't a clear path to emitting, say, multiple metrics from an input log event
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:105]: If we _don't_ want to introduce a new transform, then I'd probably lean towards updating the `remap` transform with a new option `explode = true`. If set, the transform would check at boot that your VRL script returns an array, and it would behave the same as what I described above, except that it doesn't require a separate transform and allows you to mutate the event before you explode it into zero or more events.

I still lean towards a separate transform being the best option, but this would probably be my second preference.

I realize that both of my proposals are _less flexible_ than an `emit_log`, but I think overall they come out on top due to their simplicity over the current proposal in the RFC.

[the remark this answers is not in the document]
Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:105]: Thanks for the alternate proposal!

I like the idea of a separate transform where it evaluates the last expression (similar to conditions) to get the output. This should avoid the need to decouple `.` from the underlying event for now.

> I still somewhat prefer a separate transform for this. I worry that — in a couple of months — we'll be at a point where the remap transform is a "do anything you want" transform, which it was never meant to be.

Agreed, I can see that this is expanding the scope significantly. I think I do see a future where remap is more general purpose, similar to Lua, to offer increased flexibility, but we can defer that for now until we have more data from users about what they are trying to achieve.

> > The biggest open question for me with this approach is whether the event in this transform should be _immutable_, to enforce people to use `remap` to mutate events, and `explode` to convert to multiple events, instead of using `explode` to do both.
>
> I'd probably lean to **yes**, make the event immutable, if nothing else, it gives us more flexibility down the line on what we want to do with this transform, and it also guides operators to use our transforms "for their intended purpose". We can always remove that restriction in the future, if it turns out to be a good idea to do so.

I also lean towards yes, it should be immutable in that modifications should not persist. I do think it should be possible to do some mapping in the `explode` transform though. For example, I think it will be common to want to map over the elements that are being exploded to add additional fields like:

Input:

```
{ "host": "foobar", "events": [{ "message": "foo" }, { "message": "bar" }] }
```

```toml
[transforms.explode]
type = "explode"
source = "map(array!(.events), |event| event.host = .host) ?? []"
```

The exact "mapping over elements" syntax will be contingent on #6031 .

I'll update the RFC to reflect this approach as I think it is a simpler start and we can always add in `emit_log` later if we find an eed.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 3 above)
Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:51]: Yeah, that's true, but we don't have a concept of structs in VRL so it would depend on the user passing a map of key/values that matches some schema.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+The following events will be output:
+
+```json
+{"message": "foo"}
+{"message": "bar"}
+```
+
+## Rationale
+
+This enhances vector with a native transform capable of an event into multiple events. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to remap
```
Comment (bruceg) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:103]:
```suggestion
This enhances vector with a native transform capable of an event into multiple events. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to remap.
```

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.
+
+This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.
+
+Using a separate `explode` transform keeps the responsibilities of the transform more clear and avoids having to refactor the `remap` transform to decouple `.` from the underlying Vector `Event` object; though we may still want to do this in the future anyway.
+
+## Open Questions
+
+1. How should we handle the case where the user returns an array where one or more elements is not an object? My instinct for now is just to log and drop them.
```
Comment (bruceg) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:175]: I can't think of anything else reasonable to do other than dropping the whole batch, and that seems excessively harsh.

**Review by bruceg — APPROVED** [review]
I'm not a big fan of the use of "explode" to mean separating data into multiple pieces, but otherwise this seems a reasonable solution given the desired constraints.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
+# RFC 6330 - 2021-04-07 - Emitting multiple log events from remap transform
+
+This RFC describes an approach for emitting multiple log events from a remap transform.
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:3]:
```suggestion
This RFC describes an approach for emitting multiple log events from an `explore` transform.
```

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
+# RFC 6330 - 2021-04-07 - Emitting multiple log events from remap transform
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:1]:
```suggestion
# RFC 6330 - 2021-04-07 - Emitting multiple log events from explode transform
```

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+This RFC describes an approach for emitting multiple log events from a remap transform.
+
+[Credit to @JeanMertz](https://github.com/timberio/vector/pull/7038#discussion_r609476064) for this revised approach.
+
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:6]:
```suggestion
```

If I had to give credits to everyone that helped me over the years, you'd be reading a novel for every PR or RFC I'd write 😄

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+Out:
+
+- Turning a log event into multiple metrics
+- Turning a metric into multiple metrics
+- Turning a metric into multiple logs
+
+## Motivation
+
+Currently the remap transform can only modify an event as it is passing through. Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:21]:
```suggestion
Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.
```

I don't think it's relevant to mention the remap transform in this case, as it is complimentary, but serves a different purpose.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).
+
+## Internal Proposal
+
+The proposal is to add an `explode` transform that makes use of remap to emit a set of events from one input event by requiring the remap script to resolve to an array. For each element of the array, a separate event will be published.
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:27]:
```suggestion
The proposal is to add an `explode` transform that makes use of Vectir Remap Language (VRL) to emit a set of events from one input event by requiring the VRL program to resolve to an array. For each element of the array, a separate event will be published.
```

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+type = "explode"
+source = "map(array!(.events), |event| event.host = .host) ?? []"
+```
+
+Output:
+
+```json
+{"host": "foobar", "message": "foo"}
+{"host": "foobar", "message": "bar"}
+```
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:71]: Is this relevant to mention in this RFC? It seems orthogonal.

Also, it might actually be _more performant_ to do this in a separate `remap` transform, which can then/eventually run in parallel for multiple events.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+This is similar to the support that the current [`explode` transform PR](https://github.com/timberio/vector/pull/6545) has for merging in top-level fields when creating events from a subfield that has an array.
+
+
+## Doc-level Proposal
+
+To convert an incoming event into multiple events, the `explode` transform can be used. This transform takes a VRL script where it expects the last expression to return an array.
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:78]: Just syncing terminology used throughout our docs/code.

```suggestion
To convert an incoming event into multiple events, the `explode` transform can be used. This transform takes a VRL program where it expects the last expression to return an array.
```

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 6 above)
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:103]:
```suggestion
This enhances Vector with a native transform capable of transforming an event into multiple events. Without this, users will continue to have to use Lua or WASM to achieve this, which introduces a performance bottleneck compared to a native transform.
```

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
+- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)
+
+These are similar to the proposed approach.
+
+## Drawbacks
+
+- Adds an additional transform to be aware of.
+- Less flexible than `emit_log` alternative which could emit arbitrary events from `remap` transform.
+- This idea of the last line of the remap script being the "return value" could be a bit confusing to users in how it differs from `remap` which uses whatever `.` is at the end of the script.
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:116]: This is true, but it's worth pointing out that it _is_ similar to how VRL-based conditions work (in unit-tests, but also filter and route transforms).

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 7 above)
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:175]: My proposal would be to handle types as follows:

- Objects are turned into individual events.
- Any other type is turned into an individual event, with the value of the type set as the `message`.
- We can make this configurable through a transform config if we so desire.

E.g.

```json
{ "events": [{ "timestamp": "...", "message": "foo" }, "bar", 42.0] }
```

Turns into:

```jsonl
{ "message": "foo", "timestamp": "..." }
{ "message": "bar" }
{ "message": 42.0 }
```

If operators want, they can either use a filter transform or a remap transform to shape events to their liking after this happens (e.g. enforce message becomes a string in the case of the last one, or filter out events without a timestamp in the case of the second and third one).

**Review by JeanMertz — APPROVED** [review]
I like this proposal. I left a few comments and proposed a solution to the open question, but otherwise this looks good.

@bruceg am I understanding you correctly that you're not a fan of the _name_ "explode" for a feature like this, or is there something else that bothers you about it? I'm curious if there are other names we can suggest and/or if there is precedence for the name?

**Comment by JeanMertz** [comment]
One more thing I considered while thinking this over, is that it still might be sensible to make it easier to use this transform by providing a field name instead of a VRL program.

That is, you'd write this:

```toml
[transform.explode]
type = "explode"
field = "events"
```

Which gets turned into this:

```toml
[transform.explode]
type = "explode"
script = "array!(.events) ?? []"
```

This way, for simple use-cases, people don't need to learn how to write VRL, but once they have a need for it, they swap out the `field` config for `script`, and use VRL for whatever advanced logic they need.

This could either be a "future improvement", or something we want to add straight away.

**Comment by bruceg** [comment]
> @bruceg am I understanding you correctly that you're not a fan of the _name_ "explode" for a feature like this, or is there something else that bothers you about it? I'm curious if there are other names we can suggest and/or if there is precedence for the name?

Correct, I dislike the name, not any of the other substance. PHP called its string division operation "explode", but it was aliased with "split" which makes more sense, particularly when you consider the process of reversing the operation (ie "join" not "implode"). Some other possibilities: split, bifurcate, divide, subdivide, separate, fork, part, partition, segregate. Yes, some of those are awful too, just putting it out there.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 13 above)
Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:71]: True, but how would they get the fields from the original source event in the later remap transform? I think they'd need to do the mapping in the `explode` transform.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 7 above)
Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:175]: 👍 I like this.

**Comment by jszwedko** [comment]
Thanks for the feedback both! I think I made all of the adjustments minus a couple:

* I agree `explode` isn't a great name. I would have liked to use `split` but we already have a transform named that that. Maybe I'll think of a better name over the weekend. 



> One more thing I considered while thinking this over, is that it still might be sensible to make it easier to use this transform by providing a field name instead of a VRL program.
> 
> That is, you'd write this:
> 
> ```toml
> [transform.explode]
> type = "explode"
> field = "events"
> ```
> 
> Which gets turned into this:
> 
> ```toml
> [transform.explode]
> type = "explode"
> script = "array!(.events) ?? []"
> ```
> 
> This way, for simple use-cases, people don't need to learn how to write VRL, but once they have a need for it, they swap out the `field` config for `script`, and use VRL for whatever advanced logic they need.
> 
> This could either be a potential "future improvement", or something we want to add straight away.

@JeanMertz what would you think of just loosening the compile-time check for an array result? That would let users just write:

```
[transform.explode]
type = "explode"
source = ".events"
```

We would need to error and drop values that are not arrays, but I think we'd need to do the same with your shorthand suggestion.

As another consideration, the original `explode` PR, https://github.com/timberio/vector/pull/6545, has support for something like `field` with the additional nicety of merging in top-level fields into the child event.

I do think users will want to map certain top-level "metadata" fields (like `.timestamp, `.host`, etc.) onto each created event. I'm realizing this is not super straight-forward with the proposal here. We don't yet have mapping / iteration syntax, but even when we do, we might like to make this default use-case simpler than requiring users to map the fields manually. I'll think about this some more, but it is making me think that maybe the existing `explode` transform PR has some advantages despite its inflexibility.

**Comment by binarylogic** [comment]
> One more thing I considered while thinking this over, is that it still might be sensible to make it easier to use this transform by providing a field name instead of a VRL program.

Big 👍 from me. This should be the first version. Using VRL as the only option seems a little offputting for simple cases.

**Comment by FungusHumungus** [comment]
I think you mentioned at one point that this transform should copy the top level fields into each emitted event, but I'm not seeing that discussed here.

Should:

 ```
{ hostname: "blah", events: [ { thing: "one" }, { thing: "two" } ] }
```

be exploded as:
```
{ hostname: "blah", thing: "one" }
{ hostname: "blah", thing: "two" }
```
? 

If so, which I could see as being useful, we would need to decide what happens should the fields already be in the rows. Perhaps this could be an additional option for the transform?

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+These are similar to the proposed approach.
+
+## Drawbacks
+
+- Adds an additional transform to be aware of.
+- Less flexible than `emit_log` alternative which could emit arbitrary events from `remap` transform.
+- This idea of the last line of the remap script being the "return value" could be a bit confusing to users in how it differs from `remap` which uses whatever `.` is at the end of the script. However, this is similar to VRL's use in unit tests and conditions for the `filter` and `route` transforms.
+- Ongoing maintenance burden should be minimal.
+
+## Alternatives
```
Comment (lukesteensen) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:119]: One other alternative could be a config option on the remap transform like `unnest_field = "events"` that works vaguely like SQL's `UNNEST` operator and is applied after the script is run. 

We built remap to reduce the amount that users had to chain together primitive transforms to do what they want, so this might be a little bit more in that spirit than adding a new primitive transform.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.
+
+### Modifying remap to accept setting the root object to an array
+
+https://github.com/timberio/vector/issues/6988
+
+This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.
+
+This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.
```
Comment (lukesteensen) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:169]: Would this still be difficult if Vector's `Event` could be an array at the top level?

**Comment by jszwedko** [comment]
Alright, take three here 😄 

Thanks for all of the discussion and feedback so far!

I pushed an updated proposal that:

* Goes with the original thought that `remap` could just let `.` be an array for emitting multiple events. This will take a bit of work to decouple `.` from `LogEvent`, but hopefully not too much.
* Adds a new `unnest` VRL function (thanks to Luke for the naming idea) that, given an object and a field name, will return an array of objects using the elements in the array field.
* Adds an `unnest` transform that is essentially just syntax sugar over the equivalent VRL program (I think it'd just be replaced by a `remap` transform at config build time)

@lukesteensen @JeanMertz @binarylogic @StephenWakely @bruceg do you mind taking a look at the new proposal?

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+
+Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).
+
+## Internal Proposal
+
+The proposal is to:
+
+1. Extend `remap` so that if `.` is an array at the end, it will emit one event for each element in that array
+2. Add an `unnest` VRL function that will transform an object into an array of objects using a specified field on the input object
+3. Add an `unnest` transform that can be used for simple cases (could just "expand" into a `remap` transform)
```
Comment (binarylogic) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:29]: I would vote against a separate transform just for consistency and simplicity when browsing transforms.

Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:29]: I tend to agree. I added it just based on earlier feedback on this proposal, but I'm inclined to leave it out. We can always add it later if the need arises.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+
+This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.
+
+This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.
+
+Using a separate `explode` transform keeps the responsibilities of the transform more clear and avoids having to refactor the `remap` transform to decouple `.` from the underlying Vector `Event` object; though we may still want to do this in the future anyway.
+
+## Open Questions
+
+1. Do we want the additional `unnest` transform? I included it based on [feedback that using VRL was offputting for simple cases](https://github.com/timberio/vector/pull/7038#issuecomment-817043491), but maybe the newly proposed `remap` modifications are easy enough to use that we don't need the additional transform.
```
Comment (binarylogic) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:256]: Ha, the current Ben disagrees with the past Ben. My stance has changed. Assigning `.` to an array has always struck me as the simplest approach.

**Review by binarylogic — APPROVED** [review]
I like this new version (minus the `unnest` transform). Nice work.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+{ "timestamp": "2020-12-09T16:09:53+00:00", "host": "localhost", "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Remap transform config:
+
+```toml
+[transforms.remap]
+type = "remap"
+source = """
+. = unnest(., "events", only_fields: ["host"])
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:78]: This is the first time (I believe) where we use a string to represent a path. Is there a reason why you chose this over a path-type?, e.g.:

```suggestion
. = unnest(., .events, only_fields: [.host])
```

This would bring it more in-line with functions like `del`, which also takes a path, but doesn't expand that path to a value, instead using it to manipulate the target event.

There are edge-cases to consider here (e.g. what happens when you provide an indexed-path), but overall it does seem to me like the better choice.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+
+```toml
+[transforms.unnest]
+type = "unnest"
+field = "events"
+```
+
+Which would have the same effect. I think the implementation could just take advantage of the "expand" feature of transforms that lets them replace themselves at vector start time with another transform.
+
+Additionally, we will provide `only_fields` and `except_fields` as options on the `unnest` function and transform to allow users to select which fields will be kept. These match similar semantics to the `encoding` options on sinks.
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:64]: Is there a big need for this, compared to using `del` before calling `unnest`? I guess these arguments could be considered an easier solution to achieve the same, but I'm not convinced it's worth the added complexity.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 18 above)
Comment (JeanMertz) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:29]: We can leave it off for now. I do agree if we go with the `.` solution the added transform is less of a requirement.

If we went with the previous proposal of using the final value of the program to determine the events triggered, then we'd need a new transform, since that's how how `remap` currently works.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
+- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)
+
+These are similar to the proposed approach.
+
+## Drawbacks
+
+- Adds an additional transform to be aware of.
+- Less flexible than `emit_log` alternative which could emit arbitrary events from `remap` transform.
+- Ongoing maintenance burden should be minimal.
```
Comment (JeanMertz) [inline rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:146]: An additional drawback of this solution as opposed to the previous one (separate transform, using final value of program), is that it _can_ result in unexpected outcomes if someone is doing something like this:

```jsonl
// most common events
{ "message": "{ \"foo\": true }" }

// uncommon, unexpected
{ "message": "[true, false]" }
```

```swift
. = parse_json!(.message)
```

output:

```jsonl
// expected
{ "foo": true }
```

```jsonl
// unexpected
{ "message": true }
{ "message": false }
```

By having exploding/unnesting be done in a dedicated transform, people explicitly opt-in to this behaviour, which can still cause unexpected results for badly formatted events, but it means (1) that when you see multiple events you know exactly where to debug, and (2) when using the `remap` transform, you never have to worry about the above situation occurring.

This is why I'm still in favour of the previous proposal from a useability perspective, but I'm not _opposed_ to the current approach, so I'm happy to see this land if others prefer this over the alternative.

Comment (JeanMertz) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:146]: I should note that the above example is something I've seen happen plenty of times as I worked on similar observability systems in the past (e.g. tens/hundreds of services, managed by different teams, using their own techniques/libraries to emit events, some untyped, and accidentally emitting non-conforming event shapes).

**Review by StephenWakely — APPROVED** [review]
Yeah, this is nice.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 20 above)
Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:78]: Ah, yes, I think I like that. I had actually forgotten we could just take the path without resolving it.

We could even just do:

```
. = unnest(.events)
```

Where it would just pull `events` up one level since the path sufficient information about the parent.

This does reduce the user's ability to dynamically set the parameter based on some other value though (e.g. using a different value for `only_fields` by setting a variable prior).

I could see users being slightly confused that:

```
foo = [.host]
. = unnest(., .events, only_fields = foo)
```

Doesn't work. `delete` has this same behavior though.

I'll think about this a bit more.

It is now occurring to me that `del` is also a bit special since it modifies the argument by reference. I think that is the only function that does that.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 21 above)
Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:64]: Yeah, it's not a big deal, users could workaround this via calling `del` beforehand. It's just a shortcut.

Though, we don't have an easy way to say "delete all fields _but_ these" yet.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 22 above)
Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:146]: 🤔 That is a good observation.

I think generally people neglect that `parse_json` can return other things than just objects. I think maybe we could improve the docs for that function as a start to have examples of enforcing the expected type through the type assertion functions.

I agree a separate transform would make it more obvious, but I also like the idea of users just being able to use remap.

Comment (pablosichert) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:146]: One approach to address this could be adding an explicit flag (e.g. `emit_multiple`) to the transform config, which would error when anything other than an array of objects is present in the VRL root after transformation.

Or vice versa error when an array is present but the flag was not set.

File: rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md (hunk 20 above)
Comment (jszwedko) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:68]: We can discuss this more in the PR for `unnest`, but I'm inclined to just stick with strings rather than requiring paths. I think this'll be less surprising to users.

Comment (JeanMertz) [reply rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md:68]: > It is now occurring to me that `del` is also a bit special since it modifies the argument by reference. I think that is the only function that does that.

That is correct. This was discussed during implementation. There were a few more that directly modified the target (such as `only_fields`), but that one was removed before the 1.0 release.

Ergonomically, it felt more useful to have `del` behave this way, and return the value of the deleted field, as it made renaming fields easier to do (`.foo = del(.bar)`).

> We can discuss this more in the PR for `unnest`, but I'm inclined to just stick with strings rather than requiring paths. I think this'll be less surprising to users.

I agree, we don't need to decide here, but I don't think we should use paths, or at least if we do we should be consistent and make this part of the specification (https://github.com/timberio/vector/issues/3740).
